File: python/src/algorithm/geometry/solids/rounds/cylinder.py
import math
from typing import final

from algorithm.geometry.solids import Round
from algorithm.utils import validation


@final
class Cylinder(Round):
    """
    Cylinder
    ========

    A tube-shaped solid with two parallel circular bases and one curved lateral
    surface. Examples include soda cans and pipes.

    Example
    -------
    ```
    cylinder = Cylinder(radius=5, height=10)
    print(cylinder.volume())
    ```
    """

    def __init__(self, radius: float, height: float):
        validation.validate_positive(radius, "radius")
        validation.validate_positive(height, "height")
        self._radius = radius
        self._height = height

    @property
    def radius(self) -> float:
        """
        Returns the radius of the cylinder.
        """
        return self._radius

    @property
    def height(self) -> float:
        """
        Returns the height of the cylinder.
        """
        return self._height

    @property
    def diameter(self) -> float:
        """
        Returns the diameter of the cylinder.
        """
        return self._radius * 2

    def surface_area(self) -> float:
        return 2 * math.pi * self.radius * (self.radius + self.height)

    def volume(self) -> float:
        return math.pi * pow(self.radius, 2) * self.height

File: java/src/main/java/algorithm/geometry/solids/Cylinder.java
package algorithm.geometry.solids;

import algorithm.geometry.Solid;
import algorithm.utils.Validation;

public final class Cylinder implements Solid {
    private double radius;
    private double height;

    public Cylinder(double radius, double height) throws Exception {
        Validation.validatePositive(radius, "radius");
        Validation.validatePositive(height, "height");
        this.radius = radius;
        this.height = height;
    }

    public double getRadius() {
        return radius;
    }

    public double getHeight() {
        return height;
    }

    @Override
    public double volume() {
        return Math.PI * Math.pow(radius, 2) * height;
    }

    @Override
    public double surfaceArea() {
        return 2 * Math.PI * radius * (radius + height);
    }
}
